src/Float16.h

```diff
@@ -124,4 +129,127 @@ HALIDE_ALWAYS_INLINE halide_type_t halide_type_of<Halide::float16_t>() {
     return halide_type_t(halide_type_float, 16);
 }
 
+namespace Halide {
+
+/** Class that provides a type that implements half precision
+ *  floating point using the bfloat16 format.
+ *
+ *  This type is enforced to be 16-bits wide and maintains no state
+ *  other than the raw bits so that it can passed to code that checks
+ *  a type's size and used for buffer_t allocation. */
+struct bfloat16_t {
+
+    static const int mantissa_bits = 7;
+    static const uint16_t sign_mask = 0x8000;
+    static const uint16_t exponent_mask = 0x7f80;
+    static const uint16_t mantissa_mask = 0x007f;
+
+    /// \name Constructors
+    /// @{
+
+    /** Construct from a float, double, or int using
+     * round-to-nearest-ties-to-even. Out-of-range values become +/-
+     * infinity.
+     */
+    // @{
+    bfloat16_t(float value);
+    bfloat16_t(double value);
+    bfloat16_t(int value);
+    // @}
+
+    /** Construct a bfloat16_t with the bits initialised to 0. This represents
+     * positive zero.*/
+    bfloat16_t();
+
+    /// @}
+
+    // Use explicit to avoid accidently raising the precision
+    /** Cast to float */
+    explicit operator float() const;
+    /** Cast to double */
+    explicit operator double() const;
+    /** Cast to int */
+    explicit operator int() const;
+
+    /** \name Convenience "constructors"
+     */
+    /**@{*/
+
+    /** Get a new bfloat16_t that represents zero
+     * \param positive if true then returns positive zero otherwise returns
+     *        negative zero.
+     */
+    static bfloat16_t make_zero(bool positive);
```

Comment:
Nit: reading `make_zero(true)` at a call site isn't very obvious as to what the boolean means. Maybe not used enough to matter, but something like `enum { kPositive, kNegative }` would be more obvious.